File: frontend/src/api/alerts.ts
import { apiFetch } from './client';

export type AlertType =
  | 'FLOOD_ONSET'
  | 'FLOOD_SEVERITY_ESCALATION'
  | 'HIGH_SEVERE_FLOOD_RISK'
  | 'TRAVEL_WINDOW_CLOSING'
  | 'ROUTE_AVOID'
  | 'CRITICAL_ACCESS_THREAT'
  | 'CRITICAL_ACCESS_LOSS'
  | 'PROTECT_CITY_PRIORITY'
  | 'GROUND_TRUTH_CONFLICT'
  | 'MODEL_INPUT_DEGRADED'
  | 'MODEL_UNCERTAINTY'
  | 'SIMULATOR_SCENARIO_RESULT'
  | 'SYSTEM_DATA_QUALITY';

export type AlertSeverity = 'INFO' | 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export type AlertStatus =
  | 'ACTIVE'
  | 'ACKNOWLEDGED'
  | 'RESOLVED'
  | 'EXPIRED'
  | 'SUPPRESSED'
  | 'FAILED';

export type EvidenceStrength = 'UNVERIFIED' | 'CORROBORATED' | 'CONFIRMED';

export interface AlertGeneratePayload {
  digital_twin_run_id?: string;
  travel_window_run_id?: string;
  critical_access_run_id?: string;
  protect_city_run_id?: string;
  ground_truth_run_id?: string;
  simulator_run_id?: string;
}

export interface Alert {
  alert_id: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  status: AlertStatus;
  title: string;
  summary: string;
  affected_entity_type: string;
  affected_entity_id: string;
  condition_key: string;
  fingerprint: string;
  source_phase: string;
  source_run_id: string;
  sources: Record<string, string>;
  input_completeness: number;
  model_status: string;
  evidence_strength: EvidenceStrength;
  uncertainty_status: string;
  recommended_action: string;
  governance_notice: string;
  configuration_version: string;
  generated_at: string;
  updated_at: string;
  acknowledged_at?: string;
  acknowledged_by?: string;
  resolved_at?: string;
  resolved_by?: string;
  resolution_reason?: string;
  suppressed_at?: string;
  suppressed_by?: string;
  suppression_reason?: string;
  expires_at?: string;
}

export interface AlertListResponse {
  total_count: number;
  alerts: Alert[];
}

export interface AlertEvidence {
  evidence_id: string;
  alert_id: string;
  source_phase: string;
  source_run_id: string;
  source_artifact_id?: string;
  evidence_type: string;
  metric: string;
  value?: number;
  units?: string;
  timestamp?: string;
  evidence_strength: EvidenceStrength;
  details: Record<string, any>;
}

export interface ExplainabilityStep {
  step_id: string;
  alert_id: string;
  sequence: number;
  category: string; // WHAT, WHY, WHEN, WHERE, HOW_CERTAIN, WHAT_SHOULD_I_DO, EVIDENCE
  statement: string;
  source_phase: string;
  source_run_id: string;
  source_metric?: string;
  source_value?: number;
  units?: string;
  slice_minutes?: number;
}

export interface AlertAuditEvent {
  event_id: string;
  alert_id: string;
  event_type: string;
  previous_status?: AlertStatus;
  new_status: AlertStatus;
  timestamp: string;
  actor_type: string;
  actor_reference: string;
  reason?: string;
  source_run_id?: string;
  event_metadata: Record<string, any>;
}

export interface AlertProvenance {
  alert_id: string;
  fingerprint: string;
  source_phase: string;
  source_run_id: string;
  sources: Record<string, string>;
  configuration_version: string;
  generated_at: string;
  provenance_hash: string;
  details: Record<string, any>;
}

export async function fetchAlerts(
  statusFilter?: AlertStatus,
  severityFilter?: AlertSeverity,
  typeFilter?: AlertType
): Promise<AlertListResponse> {
  const params = new URLSearchParams();
  if (statusFilter) params.append('status', statusFilter);
  if (severityFilter) params.append('severity', severityFilter);
  if (typeFilter) params.append('alert_type', typeFilter);

  const queryStr = params.toString() ? `?${params.toString()}` : '';
  return apiFetch<AlertListResponse>(`/api/v1/alerts${queryStr}`);
}

export async function generateAlerts(payload: AlertGeneratePayload): Promise<AlertListResponse> {
  return apiFetch<AlertListResponse>('/api/v1/alerts/generate', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export async function getAlertDetail(alertId: string): Promise<Alert> {
  return apiFetch<Alert>(`/api/v1/alerts/${alertId}`);
}

export async function acknowledgeAlert(alertId: string, actorRef: string = 'OPERATOR_PRIMARY', reason?: string): Promise<Alert> {
  return apiFetch<Alert>(`/api/v1/alerts/${alertId}/acknowledge`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ actor_reference: actorRef, reason }),
  });
}

export async function resolveAlert(alertId: string, resolutionReason: string, actorRef: string = 'OPERATOR_PRIMARY'): Promise<Alert> {
  return apiFetch<Alert>(`/api/v1/alerts/${alertId}/resolve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ actor_reference: actorRef, resolution_reason: resolutionReason }),
  });
}

export async function suppressAlert(alertId: string, suppressionReason: string, actorRef: string = 'OPERATOR_PRIMARY'): Promise<Alert> {
  return apiFetch<Alert>(`/api/v1/alerts/${alertId}/suppress`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ actor_reference: actorRef, suppression_reason: suppressionReason }),
  });
}

export async function getAlertEvidence(alertId: string): Promise<AlertEvidence[]> {
  return apiFetch<AlertEvidence[]>(`/api/v1/alerts/${alertId}/evidence`);
}

export async function getAlertExplainability(alertId: string): Promise<ExplainabilityStep[]> {
  return apiFetch<ExplainabilityStep[]>(`/api/v1/alerts/${alertId}/explainability`);
}

export async function getAlertAudit(alertId: string): Promise<AlertAuditEvent[]> {
  return apiFetch<AlertAuditEvent[]>(`/api/v1/alerts/${alertId}/audit`);
}

export async function getAlertProvenance(alertId: string): Promise<AlertProvenance> {
  return apiFetch<AlertProvenance>(`/api/v1/alerts/${alertId}/provenance`);
}

export async function getAlertConfiguration(): Promise<Record<string, any>> {
  return apiFetch<Record<string, any>>('/api/v1/alerts/configuration');
}

File: backend/app/schemas/alerts.py
"""
Phase 15 Alerts, Explainability, and Audit Pydantic Schemas.
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator


class AlertType(str, Enum):
    FLOOD_ONSET = "FLOOD_ONSET"
    FLOOD_SEVERITY_ESCALATION = "FLOOD_SEVERITY_ESCALATION"
    HIGH_SEVERE_FLOOD_RISK = "HIGH_SEVERE_FLOOD_RISK"
    TRAVEL_WINDOW_CLOSING = "TRAVEL_WINDOW_CLOSING"
    ROUTE_AVOID = "ROUTE_AVOID"
    CRITICAL_ACCESS_THREAT = "CRITICAL_ACCESS_THREAT"
    CRITICAL_ACCESS_LOSS = "CRITICAL_ACCESS_LOSS"
    PROTECT_CITY_PRIORITY = "PROTECT_CITY_PRIORITY"
    GROUND_TRUTH_CONFLICT = "GROUND_TRUTH_CONFLICT"
    MODEL_INPUT_DEGRADED = "MODEL_INPUT_DEGRADED"
    MODEL_UNCERTAINTY = "MODEL_UNCERTAINTY"
    SIMULATOR_SCENARIO_RESULT = "SIMULATOR_SCENARIO_RESULT"
    SYSTEM_DATA_QUALITY = "SYSTEM_DATA_QUALITY"


class AlertSeverity(str, Enum):
    INFO = "INFO"
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class AlertStatus(str, Enum):
    ACTIVE = "ACTIVE"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    RESOLVED = "RESOLVED"
    EXPIRED = "EXPIRED"
    SUPPRESSED = "SUPPRESSED"
    FAILED = "FAILED"


class EvidenceStrength(str, Enum):
    UNVERIFIED = "UNVERIFIED"
    CORROBORATED = "CORROBORATED"
    CONFIRMED = "CONFIRMED"


class AlertGenerateRequestSchema(BaseModel):
    digital_twin_run_id: str | None = Field(None, description="Authoritative Phase 9 Digital Twin Run ID")
    travel_window_run_id: str | None = Field(None, description="Authoritative Phase 10 Travel Window / Routing Run ID")
    critical_access_run_id: str | None = Field(None, description="Authoritative Phase 11 Critical Access Run ID")
    protect_city_run_id: str | None = Field(None, description="Authoritative Phase 12 Protect City Run ID")
    ground_truth_run_id: str | None = Field(None, description="Authoritative Phase 13 Ground Truth Verification Run ID")
    simulator_run_id: str | None = Field(None, description="Authoritative Phase 14 Simulator Run ID")

    @model_validator(mode="after")
    def validate_at_least_one_source(self) -> "AlertGenerateRequestSchema":
        sources = [
            self.digital_twin_run_id,
            self.travel_window_run_id,
            self.critical_access_run_id,
            self.protect_city_run_id,
            self.ground_truth_run_id,
            self.simulator_run_id,
        ]
        if not any(sources):
            raise ValueError("At least one valid authoritative source run_id must be provided for alert generation.")
        return self


class AlertAcknowledgeSchema(BaseModel):
    actor_reference: str = Field("OPERATOR_PRIMARY", description="Actor identifier acknowledging the alert")
    reason: str | None = Field(None, description="Optional acknowledgement rationale note")


class AlertResolveSchema(BaseModel):
    actor_reference: str = Field("OPERATOR_PRIMARY", description="Actor identifier resolving the alert")
    resolution_reason: str = Field(..., min_length=3, description="Mandatory resolution reason")


class AlertSuppressSchema(BaseModel):
    actor_reference: str = Field("OPERATOR_PRIMARY", description="Actor identifier suppressing the alert")
    suppression_reason: str = Field(..., min_length=3, description="Mandatory suppression rationale")


class AlertEvidenceResponseSchema(BaseModel):
    evidence_id: str
    alert_id: str
    source_phase: str
    source_run_id: str
    source_artifact_id: str | None = None
    evidence_type: str
    metric: str
    value: float | None = None
    units: str | None = None
    timestamp: str | None = None
    evidence_strength: EvidenceStrength
    details: dict[str, Any] = Field(default_factory=dict)


class ExplainabilityStepResponseSchema(BaseModel):
    step_id: str
    alert_id: str
    sequence: int
    category: str  # WHAT, WHY, WHEN, WHERE, HOW_CERTAIN, WHAT_SHOULD_I_DO, EVIDENCE
    statement: str
    source_phase: str
    source_run_id: str
    source_metric: str | None = None
    source_value: float | None = None
    units: str | None = None
    slice_minutes: int | None = None


class AlertAuditEventResponseSchema(BaseModel):
    event_id: str
    alert_id: str
    event_type: str
    previous_status: AlertStatus | None = None
    new_status: AlertStatus
    timestamp: str
    actor_type: str
    actor_reference: str
    reason: str | None = None
    source_run_id: str | None = None
    event_metadata: dict[str, Any] = Field(default_factory=dict)


class AlertProvenanceResponseSchema(BaseModel):
    alert_id: str
    fingerprint: str
    source_phase: str
    source_run_id: str
    sources: dict[str, Any]
    configuration_version: str
    generated_at: str
    provenance_hash: str
    details: dict[str, Any] = Field(default_factory=dict)


class AlertConfigurationSchema(BaseModel):
    configuration_id: str
    configuration_version: str
    thresholds: dict[str, Any]
    enabled: bool
    created_at: str


class AlertResponseSchema(BaseModel):
    alert_id: str
    alert_type: AlertType
    severity: AlertSeverity
    status: AlertStatus
    title: str
    summary: str
    affected_entity_type: str
    affected_entity_id: str
    condition_key: str
    fingerprint: str
    source_phase: str
    source_run_id: str
    sources: dict[str, Any]
    input_completeness: float
    model_status: str
    evidence_strength: EvidenceStrength
    uncertainty_status: str
    recommended_action: str
    governance_notice: str
    configuration_version: str
    generated_at: str
    updated_at: str
    acknowledged_at: str | None = None
    acknowledged_by: str | None = None
    resolved_at: str | None = None
    resolved_by: str | None = None
    resolution_reason: str | None = None
    suppressed_at: str | None = None
    suppressed_by: str | None = None
    suppression_reason: str | None = None
    expires_at: str | None = None


class AlertListResponseSchema(BaseModel):
    total_count: int
    alerts: list[AlertResponseSchema]
